File: 2025/src/bin/day2/day2_p2.rs
// use std::env;
use std::fs;
use counter::Counter;

fn main() {

    fn check(smallest_window: usize, max_window: usize, n_str: String) -> bool {
        let mut invalid_str: bool  = false;
        for i in smallest_window..(max_window+1) {
            let str_match: &str = &n_str[..i];
            let mut count = 0;
            for j in 0..(n_str.len()/i) {
                if n_str.len() % i != 0 {
                    continue
                }
                if str_match == &n_str[j*(i)..(j+1)*(i)] {
                    count += 1;
                }
            }
            if count == n_str.len()/i {
                invalid_str = true;
                break
            } else {

            }
        }
        invalid_str
    }

    let contents = fs::read_to_string("src/bin/day2/data.txt")
        .expect("Should have been able to read the file");

    let ranges: Vec<&str> = contents.split(',').collect();
    let mut invalid: Vec<i64> = vec![];

    for range in ranges.iter() {
        let start_end: Vec<&str> = range.split('-').collect();
        let start: i64 = start_end[0].parse::<i64>().unwrap();
        let end: i64 = start_end[1].parse::<i64>().unwrap();
        for n in start..=end {
            let n_str = n.to_string();
            
            let counts = n_str.chars().collect::<Counter<_>>();
            let min_count = counts.values().min().unwrap();
            let max_count = counts.values().max().unwrap();
            
            if n_str.len() == 1 {
                continue
            }

            if *min_count == n_str.len() {
                if counts.keys().len() == 1 {
                    invalid.push(n);
                    continue
                }
            }
            
            if *min_count > 1 {
                let smallest_window = n_str.len() / max_count;
                let max_window = n_str.len() / 2;
                let invalid_str_return = check(smallest_window, max_window, n_str);
                if invalid_str_return {
                    invalid.push(n);
                }
                continue
            }

        }
    }
    let mut sum: i64 = 0;
    for num in invalid.iter() {
        sum += num
    }
    println!("{:?}", invalid);
    println!("{}", sum)
}

File: 2024/day2/day2_p2.py
def main():
    with open("day2/data.txt", "r") as f:
        contents = f.read()

    def check(line):
        prev = prev_diff = None
        for curr in line:
            if not prev:
                prev = curr
                continue
            diff = curr - prev
            if abs(diff) < 1 or abs(diff) > 3:
                return False
            if prev_diff and ((prev_diff > 0 and diff < 0) or (prev_diff < 0 and diff > 0)):
                return False
            prev_diff, prev = diff, curr

        return True 

    res = 0
    for line in contents.splitlines():
        line = [int(num) for num in line.split()]
        if check(line):
            res += 1
            continue
        for i in range(len(line)):
            if check(line[:i]+line[i+1:]):
                valid = True
                res += 1
                break
    
    print(res)


if __name__ == "__main__":
    main()
